api/controller.go

```diff
@@ -1502,6 +1503,70 @@ func (c *Controller) ObjectsListObjectsHandler() objects.ListObjectsHandler {
 	})
 }
 
+func (c *Controller) ObjectsStageObjectHandler() objects.StageObjectHandler {
+	return objects.StageObjectHandlerFunc(func(params objects.StageObjectParams, user *models.User) middleware.Responder {
+		deps, err := c.setupRequest(user, params.HTTPRequest, []permissions.Permission{
+			{
+				Action:   permissions.WriteObjectAction,
+				Resource: permissions.ObjectArn(params.Repository, params.Path),
+			},
+		})
+		if err != nil {
+			return objects.NewStageObjectUnauthorized().WithPayload(responseErrorFrom(err))
+		}
+		deps.LogAction("stage_object")
+		cataloger := deps.Cataloger
+
+		repo, err := cataloger.GetRepository(deps.ctx, params.Repository)
+		if errors.Is(err, catalog.ErrNotFound) {
+			return objects.NewStageObjectNotFound().WithPayload(responseErrorFrom(err))
+		} else if err != nil {
+			return objects.NewStageObjectDefault(http.StatusInternalServerError).WithPayload(responseErrorFrom(err))
+		}
+
+		// write metadata
+		qk, err := block.ResolveNamespace(repo.StorageNamespace, params.Stats.PhysicalAddress)
+		if err != nil {
+			return objects.NewStageObjectDefault(http.StatusInternalServerError).WithPayload(responseErrorFrom(err))
+		}
+
+		if qk.StorageType.String() != deps.BlockAdapter.BlockstoreType() {
+			return objects.NewStageObjectBadRequest().WithPayload(
+				responseError("invalid storage type: %s: current block adapter is %s",
+					qk.StorageType.String(),
+					deps.BlockAdapter.BlockstoreType(),
+				))
+		}
+
+		writeTime := time.Now()
+		entry := catalog.DBEntry{
```

Comment:
What about `metadata`? I think it's valuable for these kinds of advance-use commands.

Reply:
Metadata apparently gets dropped when writing to staging since that field is never serialized.Will open a separate issue.

Reply:
it is not. Added metadata support to API and CLI